### PR TITLE
Add logic to support sending structured entries to the journald log

### LINKED ISSUE #43: logging: implement sending structured entries
This library currently implements `logging::journal_print` for sending non-structured entries to journald.

It should also expose a `journal_send` for sending structured entries, similar to the one above but taking an additional reference to a slice of `(String, String)` tuples.

/cc @sohankunkerkar

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Fixes #43 and enable us to use that logic in the `afterburn` repo to solve(https://github.com/coreos/afterburn/issues/395)

### CODE REVIEW

File: src/logging.rs [hunk 1]
```diff
@@ -82,16 +83,18 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
...
     }
 }
-
-/// Print a message to the journal with the given priority.
-///
+/// Send a message to the local systemd journal. 
+/// vars is a map of journald fields to values.
 /// This uses the default systemd-journald socket.
-pub fn journal_print(priority: Priority, msg: &str) -> Result<()> {
+fn journal_send(priority: Priority, msg: &str,vars: &mut HashMap<&str, &str>) -> Result<()> {
```
Comment (lucab) [inline src/logging.rs:89]: This would be a `pub fn`.

Also, no reason for `vars` to be marked as `mut`, I think.

File: src/logging.rs [hunk 2]
```diff
@@ -82,16 +83,18 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
         data.push_str(&field_payload);
     }
 }
-
-/// Print a message to the journal with the given priority.
-///
+/// Send a message to the local systemd journal.
```
Comment (lucab) [inline src/logging.rs:86]: `Send a message with structured properties to the journal.`

Also, please leave a line with just `///` between the title and the rest of the docstring, rustdoc will split the body.

File: src/logging.rs [hunk 3]
```diff
@@ -82,16 +83,18 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
...
+/// vars is a map of journald fields to values.
 /// This uses the default systemd-journald socket.
-pub fn journal_print(priority: Priority, msg: &str) -> Result<()> {
+fn journal_send(priority: Priority, msg: &str,vars: &mut HashMap<&str, &str>) -> Result<()> {
     let sock = UnixDatagram::unbound()?;
 
     let mut data = String::new();
     add_field_and_payload(&mut data, "PRIORITY", &(u8::from(priority)).to_string());
     add_field_and_payload(&mut data, "MESSAGE", msg);
+    for (k, v) in vars.iter() {
```
Comment (lucab) [inline src/logging.rs:95]: You'll need to skip `PRIORITY` and `MESSAGE` out of this iterator, and mention that in the docstring too.

**Comment by lucab** [comment]
A bunch of high-level things:
 * please do a `cargo fmt` pass on this PR
 * the CI here is currently stuck (I have a support ticket ongoing with Travis), so you'll manually have to run the [tests](https://github.com/lucab/libsystemd-rs/blob/master/.travis.yml) for the moment
 * I think that @jlebon was looking for a more generic method signature in #43

File: src/logging.rs [hunk 4]
```diff
@@ -82,16 +83,29 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
...
     }
 }
-
-/// Print a message to the journal with the given priority.
+/// Send a message with structured properties to the journal.
 ///
-/// This uses the default systemd-journald socket.
-pub fn journal_print(priority: Priority, msg: &str) -> Result<()> {
+/// vars is a map of journald fields to values.This function uses the default systemd-journald socket.
+/// Iterator may contains `PRIORITY` or `MESSAGE` fields. Instead of duplicating those fields,we'll use the ones which are already injected through journal.send(). So there's a condition in the iterator to skip those two values if present in the map.
```
Comment (jlebon) [inline src/logging.rs:89]: Minor: missing spaces after some of the periods and commas.

Comment (jlebon) [reply src/logging.rs:89]: Re. `PRIORITY` and `MESSAGE`, is that the same behaviour of the `sd_journal_send` API? If so, why not just rely on that directly? If not, why have a different behaviour here?

File: src/logging.rs [hunk 5]
```diff
@@ -82,16 +83,29 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
...
+pub fn journal_send<K, V>(
+    priority: Priority,
+    msg: &str,
+    vars: impl Iterator<Item = (K, V)>,
+) -> Result<()>
+where
+    K: AsRef<str>,
+    V: AsRef<str>,
```
Comment (jlebon) [inline src/logging.rs:97]: This is OK, though... if we want to truly match the API semantics, the value could technically just be bytes. UTF-8 is recommended, but binary is fully supported. So this could even be `AsRef<[u8]>`.

File: src/logging.rs (hunk 4 above)
Comment (lucab) [reply src/logging.rs:89]: `sd_journal_send` takes a varargs of C format strings plus arguments. This library does not link against `libsystemd.so`, so we are free to shape the API as we wish.

I personally like this API where `print` and `send` are similar, and we just take care of the two special arguments we have for priority and message.
I guess one alternative is to take only the iterator, which makes it more similar to `sd_journal_send`.

File: src/logging.rs (hunk 5 above)
Comment (lucab) [reply src/logging.rs:97]: It looks like both the C and the Go API handle strings, and even binary stuff like `MESSAGE_ID` is hex-encoded. I think that handling strings here is ok.

File: src/logging.rs (hunk 4 above)
Comment (jlebon) [reply src/logging.rs]: > This library does not link against `libsystemd.so`

Ahh wow, totally missed that part. Neat!

> I personally like this API where print and send are similar, and we just take care of the two special arguments we have for priority and message.

I think that's OK, but we should be more explicit in the docs that the arguments always override the equivalent fields. This is an ambiguity that the API avoids entirely.

Maybe rewording as something like:

> /// The `PRIORITY` or `MESSAGE` fields from the `vars` iterator are always ignored in favour of the `priority` and `message` arguments.

?

File: src/logging.rs (hunk 5 above)
Comment (jlebon) [reply src/logging.rs:97]: The protocol itself allows for any arbitrary binary data. `systemd-coredump` for example can be configured to store full coredump in the journal (which is why journal messages have a crazy limit like 768M). And in rpm-ostree, I almost made use of binary values as well in the journal when implementing https://github.com/projectatomic/rpm-ostree/pull/1813.

Comment (jlebon) [reply src/logging.rs:97]: (To be clear, this WFM as is too, just wanted to flag we weren't exposing the full capabilities of the API. We can always add a separate API later on if the need arises.)

Comment (lucab) [reply src/logging.rs:96]: Yeah, it is starting to get tricky. The C API has a further split between `_send` and `_sendv` for the string-vs-bytes topic.
I'd say we keep this signature for the moment and then iterate on a lower level helper on the side. I think in that case having split priority/message does not make sense either.

**Comment by lucab** [comment]
Travis reporting is still KO. Manually triggered a build, status is green: https://travis-ci.org/github/lucab/libsystemd-rs/builds/677601755. Merging.
